my/core/query.py

```diff
@@ -0,0 +1,386 @@
+        if order_by is None:
+            # https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.spy
+            [first_item], itrc = more_itertools.spy(itr)
+            # replace the 'itr' in the higher scope with itrc -- itr is consumed by more_itertools.spy
+            itr = itrc
+            # try to use a key, if it was supplied
+            # order_key doesn't use local state - it just tries to find the passed
+            # attribute, or default to the 'default' value. As mentioned above,
+            # best used for items with a similar structure
+            if order_key is not None:
+                order_by_chosen = _generate_order_by_func(first_item, key=order_key, default=default)
+                if order_by_chosen is None:
+                    raise QueryException(f"Error while ordering: could not find {order_key} on {first_item}")
+            elif order_value is not None:
+                itr1, itr2 = itertools.tee(itr, 2)  # expensive!!!
+                # TODO: add a kwarg to force lookup for every item? would sort of be like core.common.guess_datetime then
+                order_by_lookup: Dict[Any, OrderFunc] = {}
+
+                # need to go through a copy of the whole iterator here to
+                # pre-generate functions to support sorting mixed types
+                for obj_res in itr1:
+                    key: Any = _determine_order_by_value_key(obj_res)
+                    if key not in order_by_lookup:
+                        keyfunc: Optional[OrderFunc] = _generate_order_by_func(obj_res, where_function=order_value, default=default)
+                        if keyfunc is None:
+                            raise QueryException(f"Error while ordering: could not determine how to order {obj_res}")
+                        order_by_lookup[key] = keyfunc
+
+                # set the 'itr' (iterator in higher scope)
+                # to the copy (itertools.tee) of the iterator we haven't used yet
+                itr = itr2
+
+                # todo: cache results from above _determine_order_by_value_key call and use here somehow?
+                # would require additional state
+                # order_by_lookup[_determine_order_by_value_key(o)] returns a function which
+                # accepts o, and returns the value which sorted can use to order this by
+                order_by_chosen = lambda o: order_by_lookup[_determine_order_by_value_key(o)](o)
+
+        # run the sort, with the computed order by function
+        itr = iter(sorted(itr, key=order_by_chosen, reverse=reverse))  # type: ignore[arg-type, type-var]
+    else:
+        # if not already done in the order_by block, reverse if specified
+        if reverse:
+            itr = more_itertools.always_reversible(itr)
+
+    # apply limit argument
+    if limit is not None:
+        return itertools.islice(itr, limit)
+
+    return itr
+
+
+
```

Comment:
ideas for test:
- something involving mix of `Exception` and normal items to check that `Res` entries are handled properly
- something involving rogue entries for which predicates can't be computed -- guess what should happen in this case would depending on drop/raise parameters?

Reply:
Ah, I just realized. How items are determined as `Unsortable` is different if something is `order_key` or `order_val`. If you had a rogue item in `order_key`, it would just be classified as `Unsortable`, but in `order_val`, since it goes through and pre-generates order by funcs for each item, a rogue item would [cause an error here](https://github.com/karlicoss/HPI/pull/143/files#diff-13571aa0237710934b67bf68764a486c074521f32e7fe7fc04059e40a674d83bR398). Perhaps an error shouldn't be raised there, since `_handle_unsorted` below will split it into the unsorted iterable anyways?

The current behavior just raises the error:

https://github.com/karlicoss/HPI/pull/143/files#diff-13571aa0237710934b67bf68764a486c074521f32e7fe7fc04059e40a674d83bR543-R548

Reply:
Potential implementation:

```diff
diff --git a/my/core/query.py b/my/core/query.py
index 048b0ed..0583b86 100644
--- a/my/core/query.py
+++ b/my/core/query.py
@@ -70,7 +70,8 @@ def _generate_order_by_func(
         obj_res: Res[T],
         key: Optional[str] = None,
         where_function: Optional[Where] = None,
-        default: Optional[U] = None
+        default: Optional[U] = None,
+        force_unsorted: bool = False,
 ) -> Optional[OrderFunc]:
     """
     Accepts an object Res[T] (Instance of some class or Exception)
@@ -156,8 +157,12 @@ pass 'drop_exceptions' to ignore exceptions""")
         # warn here? it seems like you typically wouldn't want to just set the order by to
         # the same value everywhere, but maybe you did this on purpose?
         return lambda _o: default
-
-    return None  # couldn't compute a OrderFunc for this class/instance
+    elif force_unsorted:
+        # generate a dummy function which returns None
+        # this causes this type of object to be classified as an unsortable item
+        return lambda _o: None
+    else:
+        return None  # couldn't compute a OrderFunc for this class/instance
 
 
 def _drop_exceptions(itr: Iterator[ET]) -> Iterator[T]:
@@ -393,9 +398,9 @@ Your 'src' may have been empty of the 'where' clause filtered the iterable to no
                 for obj_res in itr1:
                     key: Any = _determine_order_by_value_key(obj_res)
                     if key not in order_by_lookup:
-                        keyfunc: Optional[OrderFunc] = _generate_order_by_func(obj_res, where_function=order_value, default=default)
-                        if keyfunc is None:
-                            raise QueryException(f"Error while ordering: could not determine how to order {obj_res}")
+                        keyfunc: Optional[OrderFunc] = _generate_order_by_func(obj_res, where_function=order_value, default=default, force_unsorted=True)
+                        # should never be none, as we have force_unsorted=True
+                        assert keyfunc is not None
                         order_by_lookup[key] = keyfunc
 
                 # set the 'itr' (iterator in higher scope)
@@ -542,10 +547,10 @@ def test_order_key_multi_type() -> None:
 
 def test_couldnt_determine_order() -> None:
 
-    import pytest
-
-    with pytest.raises(QueryException, match=r"could not determine how to order"):
-        res = list(select(iter([object()]), order_value=lambda o: isinstance(o, datetime)))
+    res = list(select(iter([object()]), order_value=lambda o: isinstance(o, datetime)))
+    assert len(res) == 1
+    assert isinstance(res[0], Unsortable)
+    assert type(res[0].obj) == object
 
 
 # same value type, different keys, with clashing keys
 ```

Reply:
Added a commit with a slightly modified version of above (just changed the names/added more documentation)